### ISSUE TITLE
Bug?  initial conditions in statespace_local_linear_trend.ipynb

### ISSUE THREAD

**Issue by josef-pkt** [statsmodels/statsmodels#2326]
The plot and fittedvalues show zero for the first observation. `llf_obs` is much larger for the first two observations.

Are these a consequence of default initialization of the statespace model?
Are they, or can the initial observation be ignored in the calculation of the Hessian?

Hessian is not positive definite and bse and rest shows nans in summary table

```
In [22]:mod.endog, res.fittedvalues(), res.llf_obs
Out[22]:
(array([[ 6.96129605,  7.04141166,  7.05272105,  6.9902565 ,  6.76272951,
          6.8134446 ,  6.68959927,  6.56385553,  6.41345896,  6.47697236,
          6.31173481,  6.31896811,  6.34388043,  6.4035742 ,  6.29341928,
          6.29341928,  6.41673228,  6.36475076,  6.48157713,  6.59850903,
          6.47543272,  6.44888939,  6.39859493,  6.18208491,  6.1737861 ,
          6.08904488,  6.00141488,  6.08221891,  5.99146455,  6.06610809,
          5.98141421,  6.07073773,  6.02827852,  5.93753621]]),
 array([[ 0.        ,  6.96129602,  7.1215273 ,  7.10650987,  7.01524931,
          6.73584183,  6.72001095,  6.61683616,  6.47777271,  6.30642515,
          6.36428315,  6.24584327,  6.24318122,  6.29795578,  6.39614529,
          6.30939695,  6.27742018,  6.4070299 ,  6.39333111,  6.50453335,
          6.65531308,  6.54603077,  6.45799668,  6.37394338,  6.12442775,
          6.06270173,  5.99480849,  5.91557807,  6.01160178,  5.96905913,
          6.04539687,  5.9846274 ,  6.06166943,  6.04229634]]),
 array([-7.82671804, -7.82669382,  0.90475446,  0.73236801, -1.53847562,
         1.06730176,  1.30591105,  1.21809647,  1.15615694, -0.00851262,
         1.22045091,  1.09972596,  0.87594676,  0.82856237,  0.85669728,
         1.33745943,  0.44326465,  1.2659223 ,  0.98580229,  0.93705686,
        -0.16130164,  0.90880972,  1.18463664, -0.36919037,  1.23563485,
         1.31697859,  1.34734057,  0.05289226,  1.33044589,  0.90964764,
         1.15824796,  1.00318757,  1.29732337,  0.83699252]))
```

**Comment by josef-pkt** [comment]
found while converting notebook to python 3, and checking math rendering.

@chadfulton ?

**Comment by ChadFulton** [comment]
> The plot and fittedvalues show zero for the first observation. 

This is expected, insofar as the fitted values are based on the filtered states, so they are:
`E[obs_t | time t information] = Z * E[state_t | time t information]`, and with diffuse initialization, `E[state_1] = 0`, so `E[obs_1 | time t information] = Z * 0 = 0`.

When I get around to adding support for the smoother, we will be able to also provide fitted values based on the smoothed states, like `E[obs_t | time T information]` (where T is all of the data), and in that case, it will be a much better estimate (i.e. not equal to zero!)

> `llf_obs` is much larger for the first two observations.
> Are these a consequence of default initialization of the statespace model?

This is a result of the approximate diffuse initialization. Often with diffuse initialization, ignoring the first `k_states` llf values is done (i.e. `loglikelihood_burn=self.k_states`).

Also something I would like to fix at some point, by adding exact diffuse initialization.

> Are they, or can the initial observation be ignored in the calculation of the Hessian?

If `loglikelihood_burn` is set, then those observations are ignored in the calculation of the Hessian.

> Hessian is not positive definite and bse and rest shows nans in summary table

The Hessian calculation definitely needs to be improved for state space models, as it is a bit unstable. In this model it may be due to the fact that I included the stochastic trend (even though it doesn't appear to be present). The `cov_params_delta` do work in this case (but they aren't always good either).

Here are the ways in which the Hessian problems will be (eventually) fixed:

1. Add support for outer product of gradients (this is the default used by Stata, and is guaranteed to be PD)
2. Add support for semi-analytic gradient / Hessian calculation: first, there is a semi-analytic way to calculate the gradient that I need to implement that is potentially superior to just using `approx_fprime_cs`. Second, *given* this new gradient, the Hessian can be calculated directly (without resorting to `approx_hess_cs`. I believe this should help substantially for OIM std. errors.
3. Finally, I would like, at some point, to add support for calculation and/or using the true analytic gradient. This requires running the Kalman smoother as well, and is a bit more computationally intensive than using numerical gradients, at least for smallish parameter sets. But once we have it, we can also use it to calculate (I think) the true analytic Hessian.

Typically for small-dimensional parameter spaces, there should be no problem using numerical gradients, after we implement (2) to fix the Hessians. For large dimensional parameter spaces, the analytic Hessian is actually faster.

(The Hessians are not generally *wrong* in the current implementation, but they are unstable, and that's the main problem)

**Comment by josef-pkt** [comment]
I roughly understand your answers. But I'm still a bit surprised that the Hessian calculation is so bad in this simple example.

- Is there an identification problem because both state equations are random walks? i.e. can the two variances really be estimated.
- I get positive bse (positive diagonal elements of the hessian) if I make mu mean reverting `self['transition'] = np.array([[0.8, 1], [0, 1]])` but the eigenvalues of the hessian are still strange (2 neg, 1 pos))

about initial conditions:

Is it possible to use an "informative prior" or warm starting on a previous estimate of the mean?
Another possibility: can the initial state mu_0 be estimated as part of MLE or is that too much of a change and not commonly used in this area?

Bonus question (not directly related to this issue):
To make the model into random slope coefficients (y = x beta_t + eps, beta_t = beta_t-1 + ....)
can we just change the `self['design'] = np.array([1., 0])`  to something like self['design'] = exog = x augmented with the `0`?

How do we get a constant if we don't have the random walk:   (mu_t - mu_bar) = alpha * (mu_{t-1} - mu_bar) + eps`, alpha < 1, or alpha=0 for random fluctuations?

**Comment by ChadFulton** [comment]
> I roughly understand your answers. But I'm still a bit surprised that the Hessian calculation is so bad in this simple example.

I agree - the Hessian is certainly the wart of the statespace module right now. I think I need to spend quality time working out exactly where the failures are happening, but I haven't had the time / motivation yet.

> Is there an identification problem because both state equations are random walks? i.e. can the two variances really be estimated.

I don't think identification should be a problem, the model is standard (and this model with this dataset is an example in a book, Commandeur and Koopman, 2007, except that they do not include the stochastic level).

It is essentially a deterministic trend that allows time-varying parameters. For example, if you start with `y_t = a + b t`, that can be rewritten recursively as `y_t = y_{t-1} + b`, where `y_0 = a`. To make the "level" (intercept) component time-varying, it just adds an error term, so that `y_t = y_{t-1} + b + eps_t`. To make the "trend" (slope) component time-varying, make it behave like a random walk, so you get: `y_t = y_{t-1} + b_t + eps_t` and `b_{t+1} = b_t + zeta_t`. This is the local linear trend model.

> I get positive bse (positive diagonal elements of the hessian) if I make mu mean reverting self['transition'] = np.array([[0.8, 1], [0, 1]]) but the eigenvalues of the hessian are still strange (2 neg, 1 pos))

(Side note: Can you check if you called `mod.hessian()` with the `transformed=True` argument when you got the indefinite eigenvalues? I can't see how we could get positive bse with a non-negative-definite Hessian)

However, I've also found a number of other modifications which give positive bse, and negative eigenvalues for the Hessian:

- Set `epsilon=1e-11` in `approx_hess_cs` (although other exponents don't work, e.g. `1e-10` and `1e-12` both fail). Not setting `epsilon` (letting it be whatever `approx_hess_cs` sets it at) does not work, and we still get NaNs.
- Setting `df['lff'] = np.log(df['ff'])*1e3` (although other multiplier don't work, e.g. `1e2` and `1e4` fail)
- Remove the stochastic component of the level
- Setting `loglikelihood_burn=3` (I have no idea why this would help in this case - my guess is that it's a stability issue rather than anything to do with the likelihood of the third observation).

Looking at a different dataset (Nile data, which is an example of the local level model in the Durbin and Koopman book), the stochastic component of the trend is estimated to be just about zero, and there the bse for the other two components tend to be well estimated for a variety of `epsilon=...` and multipliers scaling the dependent variable.

**Comment by ChadFulton** [comment]
> about initial conditions:

> Is it possible to use an "informative prior" or warm starting on a previous estimate of the mean?

Yes; all models need to be initialized prior to filtering, and right know there are three possibilities: `initialize_known`, `initialize_approximate_diffuse`, and `initialize_stationary`. With `initialize_known`, you can set anything you like

(this is also useful for on-line Kalman filtering. Suppose you have run the filter for observations 0, ..., 499 to get a results object `res1 = mod1.filter()`. When you get a new observation, you can just run a new filter with just one observation, where you initialize the new filter using the output from the old filter on the last observation, something like:
`mod2.initialize_known(res1.predicted_state[:, -1], res.predicted_state_cov[:, :, -1])`
I guess this is not so much help for estimating parameters, but for some uses I've heard it saves time).

> Another possibility: can the initial state mu_0 be estimated as part of MLE or is that too much of a change and not commonly used in this area?

This is definitely possible, but it requires modifying the filtering equations and the loglikelihood function (Durbin and Koopman refer to it as the augmented Kalman filter). Also, they show that doing this is equivalent to, but slower than, using the exact diffuse initialization technique, so that is the one I have planned to implement.

> Bonus question (not directly related to this issue):
To make the model into random slope coefficients (y = x beta_t + eps, beta_t = beta_t-1 + ....)
can we just change the self['design'] = np.array([1., 0]) to something like self['design'] = exog = x augmented with the 0?

Yes! To get time-varying slope coefficients, you're right that you put the exogenous matrix into the design matrix. Doing that would actually be separate from this particular model, since the local linear trend is a generalization of a deterministic time-trend rather than of fixed regression coefficients, but often you see both things incorporated together, so that your model would be something like `y_t = X beta_t + mu_t + eps_t` where `beta_t = beta_{t-1} + zeta_t` and `mu_t` is as it was defined in this notebook. Then, as you say, you would do something like:
```python
self['design'] = np.zeros((1, self.k_states, self.nobs))  # necessary to make it time-varying
self['design',:,0,:] = 1  # broadcasts along the `nobs` dimension
self['design',:,1,:] = 0
self['design',:,2:,:] = exog.T  # exog is probably (nobs x k_reg), but nobs is last dimension of design
```

Usually these types of models are referred to as unobserved components or structural time series models.

> How do we get a constant if we don't have the random walk: (mu_t - mu_bar) = alpha * (mu_{t-1} - mu_bar) + eps`, alpha < 1, or alpha=0 for random fluctuations?

If you want the constant to apply to the unobserved state, then rewrite as

    mu_t = (1 - alpha) mu_bar + alpha mu_{t-1} + eps

then you could estimate `mu_bar` as an MLE parameter, and in the updating step put in:

```python
self['state_intercept', 0] = (1 - alpha) params['...mu_bar...']
```

If you'd rather not add an MLE parameter, you could augment the state vector with a new `mu_bar` state (this would be like sort of like recursive least squares on a constant term).

-------------

If instead you want the constant to apply to the observed data, you would do something similar but use `obs_intercept` if it was an MLE parameter, or change the design matrix if it was a new state.

**Comment by josef-pkt** [comment]
just two comments: 
hessian in the changed example (0.8) looks good after adding `transformed=True`

with sigma2.level close to zero we are close to or almost at the boundary of the parameter space. I don't know how it works in this case (how sigma2.level enters the calculations), but we could almost expect to run into problems at the boundary in terms of calculations not just theoretically.

**Comment by bashtage** [comment]
@ChadFulton Does this old issue have any relevance?

**Comment by bashtage** [comment]
The apparently-degenerate first fitted values are confirmed by the maintainer to be expected diffuse-initialization behavior, not a bug. The separately-flagged Hessian/std-error robustness gaps have since been addressed by cov_type=opg/oim and exact diffuse initialization, with dedicated regression tests against R reference values (test_exact_diffuse_filtering.py). Closing as fixed/resolved.
